Initial checkin

ycombinator opened the pull request:
This PR introduces a first cut for this filter plugin.

The plugin converts string representations of computer storage sizes — for example, "123 MB" or "5.6Gb" — into their numeric value in bytes — for example, 128974848 and 700000000 respectively. It can also distinguish between bit and byte string units and perform the conversion appropriately.

This PR includes:
* [x] Code
* [x] Unit tests
* [x] Docs

ycombinator committed:
init checkin

ycombinator commented:
/cc @karenzone (I can't add you as a reviewer on the PR for some reason — probably having to do with being on the right GitHub teams).

ycombinator commented:
Re: docs, some comments from @jakelandis and @dedemorton off-PR that I'm mentioning here mostly so I don't forget them:

* [x] Make sure you look at an existing plugin to see how the file is structured and how we do IDs.
* ~[ ] You may want to call out the ES bytes data type for the field mapping so that Kibana will re-format it for automatically~

ycombinator committed:
Adding docs

ycombinator committed:
Changing meaning of "kb" (kilobytes => kilobits)

Per https://en.wikipedia.org/wiki/Kilobit

ycombinator commented:
@jakelandis said:

>  You may want to call out the ES bytes data type for the field mapping so that Kibana will re-format it for automatically

I'm not seeing this data type in ES docs. I see [`byte`](https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html) but that's something else. Could you point me in the right direction here please? Thanks.

ycombinator committed:
Ignoring Gemfile.lock

ycombinator committed:
Fixing CONTRIBUTORS

ycombinator committed:
Updating LICENSE

ycombinator committed:
Updating gemspec file

jakelandis commented on Gemfile:
```diff
@@ -0,0 +1,2 @@
+source 'https://rubygems.org'
+gemspec
```
Can you add this snippet below this: https://github.com/logstash-plugins/logstash-input-beats/blob/master/Gemfile#L5-L11

This allows for testing against different branch versions.

jakelandis replied:
Also, can you add a .travis file like this: https://github.com/logstash-plugins/logstash-input-beats/blob/master/.travis.yml

ycombinator replied:
@jakelandis Should both these changes be made under https://github.com/elastic/logstash/tree/master/lib/pluginmanager/templates as well for future runs of the plugin generator (`bin/logstash-plugin generate ...`)?

jakelandis replied:
Yes, and thank you for https://github.com/elastic/logstash/pull/9193

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,102 @@
+#
+class LogStash::Filters::Bytes < LogStash::Filters::Base
+
+  @@prefix_powers = {
```
should probably avoid global scope here in favor of a frozen const:
```
PREFIX_POWERS = {
...
}.freeze
```

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,102 @@
+    # Take out any non-numeric characters or periods (e.g. delimiters) from
+    # the number
+    # TODO: i18n
+    number.tr!('^0-9.', '')
```
Did you mean to leave the period in the expression ?

ycombinator replied:
Yes, for the decimal point. The comment above is unclear so I'll update that.

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,102 @@
+      result /= 8
+    end
+
+    # "kb" is a special case, we've been given bits
```
is there a difference in `kb` vs `Kb` ? 

is there a difference in `mb` vs `Mb` ?

ycombinator replied:
Yes, I'm treating `mb`, `MB`, and `mB` as megabytes and `Mb` as megabits

I wanted to give the same treatment for kilobytes vs. kilobits but [apparently](https://en.wikipedia.org/wiki/Kilobit) `kb` is supposed to be kilobits. Hence the special casing here.

jakelandis commented on logstash-filter-bytes.gemspec:
```diff
@@ -0,0 +1,23 @@
+Gem::Specification.new do |s|
+  s.name          = 'logstash-filter-bytes'
+  s.version       = '0.1.0'
```
Suggest to go ahead and bump to 1.0.0

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,102 @@
+
+    # Parse the source into the number part (e.g. 123),
+    # the unit prefix part (e.g. M), and the unit base part (e.g. B)
+    match = source.match(/^([0-9\,\.]*)\s*([kKmMgGtTpPeE]?)([bB]?)$/)
```
I imagine that rate as in `kb/s` is also a common pattern, may want to consider grabbing the post and placing it in a different field like <original_name>_PER as and translate `/s` to `second` ...of course that is a lot of scope creep.

ycombinator replied:
Yeah, a future PR perhaps?

jakelandis commented on DEVELOPER.md:
```diff
@@ -0,0 +1,2 @@
+# logstash-filter-bytes
+Example filter plugin. This should help bootstrap your effort to write your own filter plugin!
```
maybe remove this ?

jakelandis commented on spec/spec_helper.rb:
```diff
@@ -0,0 +1,2 @@
+# encoding: utf-8
```
you can probably just move this include to your spec for one less file to manage.

jakelandis commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,102 @@
+|40b|5
+|40KB|40960
+|40kB|40960
+|40.5KB|41472
```
nit: this 40.5 example is bit out of place...the others are very nice.

ycombinator replied:
Ah, I wanted to have an example that shows decimal usage.

jakelandis commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,102 @@
+|=======================================================================
+
+[id="plugins-{type}s-{plugin}-options"]
+==== Bytes Filter Configuration Options
```
A copy and paste example is usually SUPER nice to have.

jakelandis commented:
@ycombinator  
>I'm not seeing this data type in ES docs. I see byte but that's something else. Could you point me in the right direction here please? Thanks.

Doh! I was mixing up ES mapping types with Kibana's number based formatting options .
<img width="433" alt="image" src="https://user-images.githubusercontent.com/976291/36080621-2190028c-0f58-11e8-81ae-c18ba0ae69ba.png">

dedemorton commented on README.md:
```diff
@@ -0,0 +1,86 @@
+
+## Documentation
+
+Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).
```
This boiler plate text is no longer true. Comments in the source code are not converted into asciidoc. Is this boiler plate text defined somewhere so we can change it for developers who are contributing new LS plugins?

ycombinator replied:
Yeah, it got generated when I ran the plugin generator: https://www.elastic.co/guide/en/logstash/current/_how_to_write_a_logstash_filter_plugin.html#_use_the_plugin_generator_tool_3.

I see this text in other plugins' READMEs. Should I keep it here for consistency with other plugins or delete it so it reflects the current state of the world accurately? I think the latter but want to confirm first.

dedemorton replied:
I would change the text. You can look at the PRs I have open here: https://github.com/logstash-plugins/logstash-input-example/pull/12

ycombinator replied:
@dedemorton @jakelandis Since I generated this `README.md` using the plugin generator (`bin/logstash-plugin generate ...`), I think I should also make the suggested changes in:

* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/input-plugin/README.md#documentation
* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/filter-plugin/README.md#documentation,
* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/output-plugin/README.md#documentation, and
* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/codec-plugin/README.md#documentation

Otherwise any future plugin authors will end up with the wrong text. Correct?

dedemorton commented on README.md:
```diff
@@ -0,0 +1,86 @@
+
+#### 2.2 Run in an installed Logstash
+
+You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:
```
"2.1 method" almost sounds like you are referring to a product version rather than another section in the document.  I'd change the wording around and maybe break this into 2 sentences:

You can run your plugin in an installed Logstash instance by editing the `Gemfile` and pointing the `:path` to your local plugin development directory (see the previous section). Or you can build the gem and install it by using: ...

ycombinator replied:
This is also part of the text generated by the plugin generator.

dedemorton replied:
Hmm...then I'd leave it as-is for now.

dedemorton commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,102 @@
+
+==== Description
+
+Parse string representations of computer storage sizes such as "123 MB" or "5.6Gb" into
```
Add commas to make this easier to parse: "....of computer storage sizes, such as "123 MB" or "5.6Gb", into..."

dedemorton commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,102 @@
+  * Value type is <<string,string>>
+  * Default value is `description`
+
+Name of source field from which to parse the storage size string.
```
Maybe shorten this to improve readability: "Name of the source field that contains the storage size"

dedemorton commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,102 @@
+
+  * Value type is <<string,string>>
+
+Name of target field into which numeric value in bytes should be written.
```
Some slight changes:

"Name of the target field that will contain the storage size in bytes"

dedemorton reviewed (commented):
A few minor changes. Otherwise LGTM.

ycombinator committed:
Adding Gemfile snippet to allow testing multiple branch versions.

ycombinator committed:
Documentation changes

ycombinator committed:
Use local const instead of global var

ycombinator committed:
Clarify comment

ycombinator committed:
Bump version

ycombinator committed:
Remove indirection for import

ycombinator commented:
@jakelandis @dedemorton Thanks for the review. I've addressed all the feedback, either via changes or as replies to your comments. Ready for your 👀 again. Thanks!

ycombinator committed:
Adding travis CI config file

ycombinator committed:
Defining vars before using them :D

dedemorton reviewed (commented):
Minor change to boiler plate text. Otherwise, LGTM.

ycombinator committed:
Fixing document to have current instructions

ycombinator commented:
@jakelandis and @dedemorton Some of the changes you suggested above should probably go into the plugin generator codebase itself, so future plugin authors automatically get them when the run `bin/logstash-plugin generate ...`. For that I've put up https://github.com/elastic/logstash/pull/9193 and requested your review on it. Thanks!

ycombinator commented:
@dedemorton @jakelandis Also, this PR is ready for review again. Thanks!

andrewvc commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,100 @@
+class LogStash::Filters::Bytes < LogStash::Filters::Base
+
+  PREFIX_POWERS = {
+    'k' => 1, # 1 kilobyte = 1024 ^ 1 bytes
```
The prefixes mean different things depending on whether we're talking network data or size on disk.  See: https://en.wikipedia.org/wiki/Mebibyte .

It'd be nice to also support `MiB` and `KiB`, even though their use is a little rare.

More important would be to add an option that toggles between SI Units and powers of 2.

andrewvc replied:
By SI units, I mean `k = 1000`

ycombinator replied:
> It'd be nice to also support MiB and KiB, even though their use is a little rare.

I'd like to do this in a follow up PR, if that's okay?

> More important would be to add an option that toggles between SI Units and powers of 2.

Agreed. I will add this option to this PR.

jakelandis commented on spec/filters/bytes_spec.rb:
```diff
@@ -0,0 +1,208 @@
+
+    sample("32EB") do
+      expect(subject).to include("dest")
+      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024)
```
This is a really edge case (today)...but found this a while back with AWS EFS ... it is possible for exabytes to overflow a long when represented as bytes. Maybe we shouldn't support exabytes ?

ycombinator replied:
Sure, that sounds safer. We can not support them until someone asks for them. Will remove. Thanks.

jakelandis commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,110 @@
+[source,ruby]
+    filter {
+      bytes {
+        source => "my_bytes_string_field",
```
shouldn't have a comma here (it wont compile with it)

jakelandis commented:
I think the math is off when using a decimal point.

With the following config:
```
input { stdin {  }}    

filter {
  bytes {
        source => "message"
        target => "result"
      }
}
output {
  stdout { codec => rubydebug }
}
```
```
1.1KB
{
        "result" => 1126.4,
       "message" => "1.1KB",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:30:11.758Z
}
1.1MB
{
        "result" => 1153433.6,
       "message" => "1.1MB",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:30:19.701Z
}
```
EDIT:
Something is also funky with `1,1KB` vs. `1.1KB`
```
{
        "result" => 11264.0,
       "message" => "1,1KB",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:39:12.975Z
}
```

jakelandis commented:
```
{
        "result" => 0,
       "message" => "junk",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:36:16.797Z
}
```
Rather then a zero if not parse-able, I think adding a tag is more Logstash idiomatic 

For example the json-filter if it can not parse the JSON will tag the event : https://github.com/logstash-plugins/logstash-filter-json/blob/master/lib/logstash/filters/json.rb#L60 This allows for the differentiation of a zero value vs. a failure.

jakelandis commented:
I know in order to support bits, we need fractional bytes. , e.g. 1b = .125B , however for all other conversions it is a whole number.  

I am not sure the support for bit to byte conversion is worth the additional overhead of floating point type numbers in ES. https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html#_which_type_should_i_use  

I doubt that many things are measured in bits, and loosing support for them may simplify things and be more efficient to use long types.  

Thoughts ?

EDIT:
Or keep support and just handle them specially and round up / down to nearest whole number.

ycombinator commented:
@jakelandis Happy to drop bits for now and add in support later. As you pointed out they're probably not used much so we can wait till we get a request for them.

ycombinator commented:
@jakelandis said:

> Something is also funky with 1,1KB vs. 1.1KB
>
>```
>{
>  "result" => 11264.0,
>  "message" => "1,1KB",
>  "@version" => "1",
>  "host" => "macd",
>  "@timestamp" => 2018-02-23T19:39:12.975Z
>}
>```

I think this is expected math. The plugin doesn't currently support `,` as an alternative to `.` (the decimal point). So `1,1KB` is treated as `11KB` which converts to 11 * 1024 = 11264 bytes.

[EDIT] What we could do, preferably in a follow up PR, is to support options such as `decimal_delimiter` and `digit_group_delimiter`. Thoughts?

jakelandis commented:
@ycombinator - It looks like `.` is the minority , and `,` is the majority for decimal separators. https://en.wikipedia.org/wiki/Decimal_separator.  

I _think_ `1,1` is incorrect to assume the `,` is a group deliminator since it does not represent a hundred(th) grouping. `1,123` could be group or decimal, and this is where an explicit option would help to disambiguate . 

Is it possible to put in some small heuristics... if more then 1 `,` or `.` , it is a group deliminator, if it only 1 `,` or `.` AND the digits are not = 3 then it is  decimal deliminator...but if digits = 3 then use configurable option `decimal_delimiter` to disambiguate. 

I would consider ^^ to be close to a "must have" given our global community.  I am fine if that is a future PR, but should probably be addressed before this is a default plugin. 

Also, we should probably handle `_` and `<space>` group delimiters... i think we can do that auto-matically without exposing an explicit option assuming we expose `decimal_delimiter` to disambiguate the commas and periods for 3 based digits.

ycombinator commented:
> I think 1,1 is incorrect to assume the , is a group deliminator since it does not represent a hundred(th) grouping

Yes, of course. Duh! :)

> Is it possible to put in some small heuristics... if more then 1 , or . , it is a group deliminator, if it only 1 , or . AND the digits are not = 3 then it is decimal deliminator...but if digits = 3 then use configurable option decimal_delimiter to disambiguate.

I like this approach! Will implement.

ycombinator committed:
Fixing syntax in doc example

ycombinator committed:
Removing support for exa-

ycombinator committed:
Removing support for bits

ycombinator committed:
Using tags to indicate parse failure instead of returning 0

ycombinator committed:
Adding option for "binary" vs. "metric" prefix systems

ycombinator committed:
Adding unit test for invalid prefix system

ycombinator commented:
@jakelandis I thought about the heuristics-based approach some more and I think it might be trappy for our users (e.g. when I specify `1,333mb` do I mean `1333mb` or `1333kb`?). I'd rather give users _explicit_ options to override the [digits group separator](https://en.wikipedia.org/wiki/Decimal_separator#Digit_grouping) (defaulting to ` ` (a space)) and the [decimal separator](https://en.wikipedia.org/wiki/Decimal_separator) (defaulting to `.`). The defaults were chosen based on [current standards](https://en.wikipedia.org/wiki/Decimal_separator#Current_standards). WDYT?

ycombinator committed:
Adding options for digit group separator and decimal separator

ycombinator committed:
Updating docs

ycombinator committed:
Adding heuristics

ycombinator commented:
Never mind @jakelandis, I just re-read your heuristic more carefully and I don't think its trappy. I've implemented it in 3f4802a.

This PR is ready for your 👀 again. Thanks!

jakelandis commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,119 @@
+[cols="<,<",options="header",]
+|=======================================================================
+|Input string |Prefix system | Numeric value in bytes
+|40|`binary` or `decimal`|40
```
the doc below states that this can be `binary` or `metric` ... but the doc here says `binary` or `decimal`

ycombinator replied:
Yeah, this should be `metric`. Good catch! Fixing.

ycombinator committed:
Fixing prefix system naming inconsistency

jakelandis commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,119 @@
+Name of the target field that will contain the storage size in bytes
+
+[id="plugins-{type}s-{plugin}-prefix_system"]
+===== `prefix_system` 
```
is `prefix_system` a standard term ?  if not can i suggest `conversion_method` or the like ? _(I understand functionality here.. but still don't understand what a prefix_system is)_

ycombinator replied:
No, I don't think its standard. I made it up after reading https://en.wikipedia.org/wiki/Metric_prefix and https://en.wikipedia.org/wiki/Binary_prefix. If it doesn't make sense I'm happy to change it; `conversion_method` seems better.

jakelandis commented on docs/index.asciidoc:
```diff
@@ -0,0 +1,119 @@
+
+Which prefix system to use when converting to bytes. `binary` uses `1K = 1024B`. `metric` uses `1K = 1000B`.
+
+[id="plugins-{type}s-{plugin}-decimal_separator"]
```
nitpick: should it be decimal or fractional separator ?  is using a comma as the "decimal" separator a very American way to think about it ? (_presumes to use comma instead of the decimal_)

ycombinator replied:
I chose decimal separator because that's what Wikipedia seems to call it: https://en.wikipedia.org/wiki/Decimal_separator. I think we should leave this one as-is.

ycombinator committed:
Renaming prefix system to conversion method

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,151 @@
+  config :target, :validate => :string
+
+  # Conversion method, either "binary" (1K = 1024B) or "metric" (1K = 1000B)
+  config :conversion_method, :validate => :string, :default => "binary"
```
you can add some validation to this like here: https://github.com/logstash-plugins/logstash-input-tcp/blob/master/lib/logstash/inputs/tcp.rb#L80

ycombinator replied:
Sweet!

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,151 @@
+  config :conversion_method, :validate => :string, :default => "binary"
+
+  # Decimal separator
+  config :decimal_separator, :validate => :string, :default => "."
```
validation here too...

ycombinator committed:
Adding validations for enum options

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,151 @@
+
+    source = event.get(@source)
+
+    if !VALID_CONVERSION_METHODS.include?(@conversion_method)
```
I think you can remove this since the config is doing the validation now.

jakelandis commented:
It looks you dropped support for bits, but still output a float type:

`1.001kb -> 1025.024`

The fractional bits highlight some in-acuracy in the math w/r/t the fractional portions why doing base2 conversions.   For example, 1.001kb says 1024 bytes + 1 byte = 1025 , not 1025.024.  I think the accuracy here is acceptable, and getting the fractional accuracy for base2 is non-trival (i think). 

To avoid the mis-perception of high accuracy for fractional values, and no need for bits, can we either do whole number math, or fractional math then convert to long prior to the result ?

ycombinator commented:
> To avoid the mis-perception of high accuracy for fractional values, and no need for bits, can we either do whole number math, or fractional math then convert to long prior to the result ?

Yeah, agreed. Let me see what works but the end result will be a long.

ycombinator committed:
Rounding result to nearest whole number

ycombinator commented:
@jakelandis I decided to round the _result_ to the nearest whole number (as opposed to converting the _input_ to an integer). I think this will give better precision overall. WDYT? b8bebda

jakelandis commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,145 @@
+
+    # Parse the source into the number part (e.g. 123),
+    # the unit prefix part (e.g. M), and the unit suffix part (e.g. B)
+    match = source.match(/^([0-9#{DIGIT_GROUP_SEPARATORS}#{@decimal_separator}]*)\s*([kKmMgGtTpPeE]?)([bB]?)$/)
```
Looks like match is still supporting Exabytes which is causing errors on input like 1Eb. 

Interestingly it seems to work fine when converting to extremely large numbers (1 google gigabytes converts fine)... so maybe we should support exabytes ?  (It seems that Ruby is being kind and auto converting to numeric values that support larger then long values)

ycombinator committed:
Make exabytes supported again

ycombinator commented:
@jakelandis I added back exabytes support in 8fecb2b. Ready for review again. Thanks!

jakelandis commented:
@ycombinator  ....found one more issue, looks like the second grouping of digits is being treated as the decimal separator. 

```
filter {
  bytes {
        source => "message"
        target => "result"
        decimal_separator => "."
      }
}
```

```       
"message" => "1,000,000"
 "result" => 1000
```
```
 "message" => "1,000,999"
 "result" => 1001
```

ycombinator commented:
Thanks @jakelandis. While fixing that bug, I realized that the code is trying to be smart about when to use the `decimal_separator` option. Currently it tries to use that option _only_  if there are exactly 3 digits to the right of the rightmost separator (which could be either `.` or `,`), for disambiguation purposes. Otherwise (if there are more than or less than 3 digits to the right of the rightmost separator), the code just treats the rightmost separator as the decimal separator and everything to its left as one smooshed-together number.

Thinking about this some more, I think this attempt to be smart could actually lead to confusion. I'm thinking it might be better to _always_ use the `decimal_separator` option, defaulting it to `.` (valid choices are `.` and `,`). WDYT?

jakelandis commented:
> always use the decimal_separator

+1 I originally suggested the "smartness"...but I agree it is probably best not to try to be too smart here.  

Also, can I suggest that based on the decimal_separator, you don't include that value in the list of possible digit group separators, since you _should_ never the same value for digit and decimal separator (i _think_)

EDIT: clarified statement

ycombinator commented:
> Also, can I suggest that based on the decimal_separator, you don't include that value in the list of possible digit group separators, since you should never the same value for digit and decimal separator (i think)

So another way to think of this would be: don't allow the decimal separator to show up more than once.  Is that right? Also, what should we (Logstash) do in this case?

jakelandis commented:
> don't allow the decimal separator to show up more than once. Is that right?

yeah, that is logically equivalent

> Also, what should we (Logstash) do in this case?

tag it as a failure and optionally log why at debug.

ycombinator committed:
Always respect the decimal_separator option. No more guesswork!

ycombinator commented:
@jakelandis This is ready for review again. Thanks!

jakelandis commented:
Is this the right behavior ?
```
      decimal_separator => ","
```

```
  "message" => "1,000,000"
  "result" => 1,
```
  
I think we should error in this case (more then 1 decimal separator) in case there are is a mix of separators in the data.

ycombinator commented:
Thanks @jakelandis. I added a [check](https://github.com/logstash-plugins/logstash-filter-bytes/pull/1/commits/3f57b2cc50208fc3324a2ee87163556d01812f47#diff-a11cdef7d1b7142e0a2fdf5bb62f9bddR90) for multiple decimal separators but its happening a bit too late in the execution. I'll fix it shortly.

ycombinator committed:
Putting number of decimals check before normalizing number

ycombinator commented:
@jakelandis Fixed the location of the multiple decimal separators check in the code in f72e78e. Also added a test for your example that caught this issue.

jakelandis commented:
![perfect-1300767_640](https://user-images.githubusercontent.com/976291/39074301-67ea936c-44b6-11e8-9307-5a21ff865c1e.png)

LGTM :shipit:

robbavey commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,116 @@
+    'e' => 6, # 1 exabyte = 1024 ^ 6 bytes
+  }.freeze
+
+  DIGIT_GROUP_SEPARATORS = " _,."
```
Is there potentially a case for making this configurable too (not necessarily in this initial commit)? Digits are commonly separated by spaces in Europe, and by an apostrophe in Switzerland.

ycombinator replied:
++ to making it configurable. I'd also like to defer that to a follow up PR as this one's pretty large already.

robbavey commented on lib/logstash/filters/bytes.rb:
```diff
@@ -0,0 +1,116 @@
+
+  private
+  def normalize_number(number)
+    return number
```
Nit: Unnecessary `return` statement

robbavey commented on spec/filters/bytes_spec.rb:
```diff
@@ -0,0 +1,271 @@
+    end      
+  end
+
+  describe "non-default decimal separator" do
```
Maybe could do with a couple more tests, particularly using a digit separator as well as a decimal separator.

ycombinator replied:
I just added a test using a digit separator with a non-default decimal separator in e3bff38. There's already [a test](https://github.com/logstash-plugins/logstash-filter-bytes/pull/1/files#diff-272b5895421d51cf158a043580c1197aR221) using a digit separator with the default decimal separator.

robbavey reviewed (commented):
@ycombinator I've left a couple of small nits/comments, but nothing that should block release. Overall, this is a really nice plugin and will be a great addition.

ycombinator committed:
Removing unnecessary return statement

ycombinator committed:
Adding test using digit separator and non-default decimal separator
